sycl/include/CL/sycl/atomic_ref.hpp

```diff
@@ -0,0 +1,668 @@
+
+#if __cplusplus < 201703L
+__SYCL_INLINE_CONSTEXPR memory_order memory_order_relaxed =
+    memory_order::relaxed;
+__SYCL_INLINE_CONSTEXPR memory_order memory_order_acquire =
+    memory_order::acquire;
+__SYCL_INLINE_CONSTEXPR memory_order memory_order_release =
+    memory_order::release;
+__SYCL_INLINE_CONSTEXPR memory_order memory_order_acq_rel =
+    memory_order::acq_rel;
+__SYCL_INLINE_CONSTEXPR memory_order memory_order_seq_cst =
+    memory_order::seq_cst;
+#endif
+
+template <typename T>
+using IsValidAtomicRefType =
+    bool_constant<IsValidAtomicType<T>::value || std::is_pointer<T>::value>;
+
+// DefaultOrder parameter is limited to read-modify-write orders
+template <memory_order Order>
+using IsValidDefaultOrder = bool_constant<Order == memory_order::relaxed ||
+                                          Order == memory_order::acq_rel ||
+                                          Order == memory_order::seq_cst>;
+
+template <memory_order ReadModifyWriteOrder> struct memory_order_traits;
+
+template <> struct memory_order_traits<memory_order::relaxed> {
+  static constexpr memory_order read_order = memory_order::relaxed;
+  static constexpr memory_order write_order = memory_order::relaxed;
+};
+
+template <> struct memory_order_traits<memory_order::acq_rel> {
+  static constexpr memory_order read_order = memory_order::acquire;
+  static constexpr memory_order write_order = memory_order::release;
+};
+
+template <> struct memory_order_traits<memory_order::seq_cst> {
+  static constexpr memory_order read_order = memory_order::seq_cst;
+  static constexpr memory_order write_order = memory_order::seq_cst;
+};
+
+inline constexpr memory_order getLoadOrder(memory_order order) {
+  switch (order) {
+  case memory_order_relaxed:
+    return memory_order_relaxed;
+
+  case memory_order_acquire:
+  case memory_order::__consume_unsupported:
+  case memory_order_acq_rel:
+  case memory_order_release:
+    return memory_order_acquire;
+
+  case memory_order_seq_cst:
+    return memory_order_seq_cst;
+  }
+}
+
+template <typename T, typename = void> struct bit_equal;
+
+template <typename T>
+struct bit_equal<T, typename detail::enable_if_t<std::is_integral<T>::value>> {
+  bool operator()(const T &lhs, const T &rhs) { return lhs == rhs; }
+};
+
+template <> struct bit_equal<float> {
+  bool operator()(const float &lhs, const float &rhs) {
+    auto LhsInt = sycl::bit_cast<uint32_t>(lhs);
+    auto RhsInt = sycl::bit_cast<uint32_t>(rhs);
+    return LhsInt == RhsInt;
+  }
+};
+
+template <> struct bit_equal<double> {
+  bool operator()(const double &lhs, const double &rhs) {
+    auto LhsInt = sycl::bit_cast<uint64_t>(lhs);
+    auto RhsInt = sycl::bit_cast<uint64_t>(rhs);
+    return LhsInt == RhsInt;
+  }
+};
+
+// Functionality for any atomic of type T, reused by partial specializations
+template <typename T, memory_order DefaultOrder, memory_scope DefaultScope,
+          access::address_space AddressSpace>
```

Comment:
The SYCL 2020 version of `atomic_ref` defaults this last template argument (see https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atomic-references):
```c++
template <typename T, memory_order DefaultOrder, memory_scope DefaultScope,
          access::address_space Space = access::address_space::generic_space>
class atomic_ref;
```

I'm not sure if `access::address_space::generic_space` is defined yet, so you may need to add it.

We should probably add some new tests using atomics with the generic address space, as well.

Reply:
@Pennycook `access::address_space::generic_space` is not defined yet, I've define it but without any logic, just a variant in the enumeration. Also, `sycl::atomic_ref` was modified to use `access::address_space::generic_space` as the default value for `AddressSpace` template parameter. Please, have a look.